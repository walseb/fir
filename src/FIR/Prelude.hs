{-# LANGUAGE CPP           #-}
{-# LANGUAGE TypeOperators #-}

-- | An internal module which re-exports some definitions needed
-- from the Prelude and related modules.
--
-- As a user of the EDSL, you should never need to import this module.
module FIR.Prelude
  ( module Prelude
  , module Data.Int
  , module Data.Word
  , module Numeric.Half
  , module Data.Complex
  , module Data.Coerce
  , module Data.String
  , module Data.Kind
  )
  where

import Prelude
  ( Bool(..), Maybe(..), Either(..)
  , Float, Double
  , Semigroup(..), Monoid(..)
  , IO
  , FilePath
#if __GLASGOW_HASKELL__ >= 903
  , type (~)
#endif
  , ($), (.)
  , otherwise, maybe, either
  , fst, snd
  , curry, uncurry
  , subtract
  , id, const
  , flip
  , until, asTypeOf
  , error, errorWithoutStackTrace
  , seq, ($!)
  )

-- NB: deliberately not re-exporting Int/Word.
import Data.Int
  ( Int8, Int16, Int32, Int64 )
import Data.Word
  ( Word8, Word16, Word32, Word64 )
import Numeric.Half
  ( Half )

import Data.Complex
  ( Complex(..) )

import Data.Coerce
  ( Coercible, coerce )

import Data.String
  ( IsString(..) )

import Data.Kind
  ( Type, Constraint )
