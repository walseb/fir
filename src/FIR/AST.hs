{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE FlexibleContexts    #-}
{-# LANGUAGE FlexibleInstances   #-}
{-# LANGUAGE GADTs               #-}
{-# LANGUAGE PolyKinds           #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications    #-}
{-# LANGUAGE TypeFamilies        #-}
{-# LANGUAGE TypeOperators       #-}

{-|
Module: FIR.AST

Representation of programs using abstract syntax trees.

The user interface to the AST is through type class overloading,
see "FIR.Instances.AST" and "FIR.Instances.Codensity".

This AST uses a higher-order abstract syntax representation (HOAS),
as seen in the paper /Combining Deep and Shallow Embeddings of Domain-Specific Languages/
by Josef Svenningsson and Emil Axelsson.
-}

module FIR.AST
  ( -- * Main AST data type
    AST(..)
    -- * Syntactic type class
  , Syntactic(Internal, toAST, fromAST)
    -- * Primitive operations
  , primOp
  , HasUndefined(undefined)
    -- * Displaying ASTs graphically
  , toTree
  )
  where

-- base
import Prelude
  hiding ( undefined )
import qualified Prelude
import Data.Kind
  ( Type )
import Data.Proxy
  ( Proxy(Proxy) )
import qualified GHC.Stack
import GHC.TypeLits
  ( Symbol, KnownSymbol, symbolVal )
import GHC.TypeNats
  ( Nat, KnownNat, natVal )

-- containers
import Data.Tree
 ( Tree(Node) )

-- mtl
import Control.Monad.State.Lazy
  ( evalState )

-- tree-view
import Data.Tree.View
  ( showTree )

-- fir
import CodeGen.Instruction
  ( ID(ID) )
import CodeGen.Monad
  ( MonadFresh(fresh), runFreshSuccT )
import Control.Type.Optic
  ( Gettable, Settable, Indices )
import Control.Monad.Indexed
  ( (:=) )
import Data.Function.Variadic
  ( NatVariadic )
import Data.Product
  ( HList, IsProduct )
import Data.Type.Known
  ( Known, knownValue )
import Data.Type.List
  ( SLength )
import FIR.ASTState
  ( TLInterface
  , ASTState(..)
  )
import FIR.Binding
  ( BindingsMap
  , FunctionType, Var, Permissions
  )
import FIR.Instances.Bindings
  ( ValidDef, ValidFunDef, FunctionTypes
  , AddBinding, AddFunBinding
  , FunctionDefinitionStartState, FunctionDefinitionEndState
  , ValidEntryPoint, SetInterface
  , GetExecutionInfo
  , EntryPointStartState, EntryPointEndState
  , Embeddable
  )
import FIR.Instances.Optics
  ( User, Assigner, Viewer, Setter, KnownOptic, SOptic, showSOptic )
import FIR.Prim.Image
  ( ImageOperands )
import FIR.Prim.Op
  ( PrimOp(PrimOpType, PrimOpConstraint, opName) )
import FIR.Prim.Singletons
  ( PrimTy, primTy, KnownVars
  , PrimFunc, primFuncName
  , KnownArity
  )
import Math.Algebra.GradedSemigroup
  ( GradedSemigroup(Grade, (:<!>:)) )
import Math.Linear
  ( V, M )
import qualified SPIRV.PrimOp as SPIRV
import qualified SPIRV.PrimTy as SPIRV
import qualified SPIRV.Stage  as SPIRV
  ( ExecutionModel, ExecutionInfo )

------------------------------------------------------------
-- main AST data type

infixl 9 :$

-- | AST representation of the EDSL.
data AST :: Type -> Type where

  -- | Lambda abstraction
  Lam :: (AST a -> AST b) -> AST (a -> b)

  -- | Function application
  (:$) :: AST (a -> b) -> AST a -> AST b

  -- | Haskell-level constants can be embedded into the AST.
  Lit :: ( PrimTy a, KnownArity a ) => a -> AST a
  -- | @SPIR-V@ primitive operations
  PrimOp :: ( PrimOp op a, PrimOpConstraint op a )
         => Proxy a -> Proxy op -> AST (PrimOpType op a)

  -- Indexed monadic operations (for the AST itself)
  -- | Indexed /return/
  Return :: AST (a -> (a := i) i)
  -- | Indexed /angelic bind/
  Bind :: AST ( (a := j) i -> (a -> q j) -> q i )

  -- | Defining a new constant/variable.
  Def :: forall k ps a i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , Known Permissions ps
        , PrimTy a
        , ValidDef k i
        )
      => Proxy k  -- ^ Variable name.
      -> Proxy ps -- ^ Permissions (read,write,...).
      -> AST ( a -> (a := AddBinding k (Var ps a) i) i )

  -- | Defining a new function.
  --
  -- Meaning of type variables:
  -- * name: function name,
  -- * as: named function arguments (usually inferred),
  -- * b: function return type (usually inferred),
  -- * j_bds: bindings state at end of function definition (usually inferred),
  -- * i: monadic state at function definition site (usually inferred).
  FunDef :: forall name as b j_bds i.
            ( GHC.Stack.HasCallStack
            , KnownSymbol name
            , KnownVars as
            , PrimTy b
            , ValidFunDef name as i j_bds
            , '(as,b) ~ FunctionTypes name i
            )
         => Proxy name -- ^ Funtion name.
         -> Proxy as   -- ^ Function argument types.
         -> Proxy b    -- ^ Function return type.
         -> AST ( ( b := FunctionDefinitionEndState name as j_bds i )
                    ( FunctionDefinitionStartState name as i )
                -> ( FunctionType as b := AddFunBinding name as b i ) i
                )

  -- | Defining a new entry point.
  --
  -- An entry point is like a function definition with no arguments and Unit return type.
  -- Code within an entry point is given access to additional builtins.
  --
  -- Meaning of type variables:
  -- * name: entry point name,
  -- * stage: entry point stage,
  -- * stageInfo: entry point stage info (usually inferred),
  -- * j_bds: bindings state at end of entry point definition (usually inferred),
  -- * j_iface: entry point interface (usually inferred),
  -- * i: monadic state at entry point definition site (usually inferred),
  Entry :: forall
             ( name      :: Symbol )
             ( stage     :: SPIRV.ExecutionModel          )
             ( stageInfo :: SPIRV.ExecutionInfo Nat stage )
             ( j_bds     :: BindingsMap )
             ( j_iface   :: TLInterface )
             ( i         :: ASTState    )
           .
           ( GHC.Stack.HasCallStack
           , KnownSymbol name
           , Known SPIRV.ExecutionModel stage
           , Known (SPIRV.ExecutionInfo Nat stage) stageInfo
           , ValidEntryPoint name stageInfo i j_bds
           , stageInfo ~ GetExecutionInfo name stage i
           )
         => Proxy name      -- ^ Entry point name.
         -> Proxy stageInfo -- ^ Entry point stage info.
         -> AST (  ( () := EntryPointEndState name stageInfo j_bds j_iface i )
                   ( EntryPointStartState name stageInfo i )
                -> ( () := SetInterface name stageInfo j_iface i ) i
                )

  -- | /Use/ an optic, returning a monadic value read from the (indexed) state.
  --
  -- Like @use@ from the lens library.
  Use :: forall optic.
         ( GHC.Stack.HasCallStack, KnownOptic optic, Gettable optic )
       => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
       -> SOptic optic            -- ^ Singleton for the optic.
       -> AST ( User optic )
  -- | /Assign/ a new value with an optic.
  --
  -- Like @assign@ from the lens library.
  Assign :: forall optic.
           ( GHC.Stack.HasCallStack, KnownOptic optic, Settable optic )
         => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
         -> SOptic optic            -- ^ Singleton for the optic.
         -> AST ( Assigner optic )
  -- | /View/: access a value using an optic.
  --
  -- Like @view@ from the lens library.
  View :: forall optic.
          ( GHC.Stack.HasCallStack, KnownOptic optic, Gettable optic )
       => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
       -> SOptic optic            -- ^ Singleton for the optic.
       -> AST ( Viewer optic )
  -- | /Set/: set a value using an optic.
  --
  -- Like @set@ from the lens library.
  Set :: forall optic.
         ( GHC.Stack.HasCallStack, KnownOptic optic, Settable optic )
      => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
      -> SOptic optic            -- ^ Singleton for the optic.
      -> AST ( Setter optic )

  If    :: ( GHC.Stack.HasCallStack
           , PrimTy a
           )
        => AST ( Bool -> a -> a -> a )
  IfM   :: GHC.Stack.HasCallStack
        => AST ( Bool -> (a := j) i -> (a := k) i -> (a := i) i )
  While :: GHC.Stack.HasCallStack
        => AST ( ( Bool := i ) i -> (() := j) i -> (() := i) i )

  -- | Encapsulate local state.
  Locally :: AST ( (a := j) i -> (a := i) i )
  -- | Embed a computation into one with larger starte.
  Embed :: Embeddable i j => AST ( (a := i) i -> (a := j) j )

  -- functor/applicative operations
  Fmap :: forall f a b. ( PrimFunc f, PrimTy a, KnownArity b )
       => AST ( (a -> b) -> f a -> f b )
  Pure :: forall f a. ( PrimFunc f, KnownArity a )
       => AST ( a -> f a )
  Ap   :: forall f a b. ( PrimFunc f, PrimTy a, KnownArity b )
       => AST ( f (a -> b) -> f a -> f b )

  MkVector :: (KnownNat n, PrimTy a)
           => Proxy n
           -> Proxy a
           -> AST ( NatVariadic n a ( V n a ) )
  GradedMappend :: ( GradedSemigroup g k, a ~ Grade k g i, b ~ Grade k g j )
    => AST ( a -> b -> Grade k g (i :<!>: j) )

  -- Newtype wrapping/unwrapping.
  Mat   :: (KnownNat m, KnownNat n) => AST ( V m (V n a) -> M m n a )
  UnMat :: (KnownNat m, KnownNat n) => AST ( M m n a -> V m (V n a) )
  Ops   :: ImageOperands props ops -> AST ( ImageOperands props ops )
  Coerce :: forall a b. AST (a -> b)

  -- MkStruct
  -- MkArray

  -- | Internal HList data type.
  --
  -- Only used for providing multiple run-time indices to product optics.
  -- See [FIR issue #13](https://gitlab.com/sheaf/fir/issues/13).
  NilHList  :: AST ( HList '[] )
  ConsHList :: AST ( a -> HList as -> HList (a ': as) )
  HeadHList :: AST ( HList (a ': as) -> a )
  TailHList :: AST ( HList (a ': as) -> HList as )
  ProductToHList :: IsProduct p as => AST (p -> HList as)

  -- | Undefined.
  Undefined :: PrimTy a => AST a

  -- | As @SPIR-V@ is based around identifiers,
  -- this function can be used to create values of any type using their IDs.
  MkID :: (ID, SPIRV.PrimTy) -> AST a

------------------------------------------------

-- | Syntactic type class (Axelsson, Svenningsson).
class Syntactic a where
  type Internal a
  toAST :: a -> AST (Internal a)
  fromAST :: AST (Internal a) -> a

instance Syntactic (AST a) where
  type Internal (AST a) = a
  toAST   = id
  fromAST = id

instance (Syntactic a, Syntactic b) => Syntactic (a -> b) where
  type Internal (a -> b) = Internal a -> Internal b
  toAST        f    = Lam ( toAST . f . fromAST )
  fromAST (Lam f) a = fromAST ( f  $ toAST a )
  fromAST      f  a = fromAST ( f :$ toAST a )

------------------------------------------------

-- | Utility function for defining primops.
primOp :: forall a op r. ( PrimOp op a, PrimOpConstraint op a
                         , Syntactic r, Internal r ~ PrimOpType op a
                         )
       => r 
primOp = fromAST $ PrimOp @op @a Proxy Proxy

-- | Types at which we can define "undefined."
class HasUndefined a where
  undefined :: a

instance PrimTy a => HasUndefined (AST a) where
  undefined = Undefined

instance {-# OVERLAPPABLE #-} HasUndefined a where
  undefined = Prelude.undefined

------------------------------------------------
-- display AST for viewing

toTreeArgs :: forall m ast. MonadFresh ID m => AST ast -> [Tree String] -> m (Tree String)
toTreeArgs (f :$ a) as = do
  at <- toTreeArgs a []
  toTreeArgs f (at:as)
toTreeArgs (Lam f) as = do
  v <- fresh
  let var = MkID (v,undefined)
  body <- toTreeArgs (f var) []
  return $ case as of
    [] -> Node ("Lam " ++ show v) [body]
    _  -> Node  ":$"              (body : as)
toTreeArgs (PrimOp (_ :: Proxy a) (_ :: Proxy op) ) as 
  = return (Node ("PrimOp " ++ show ( SPIRV.op ( opName @_ @_ @op @a ) ) ) as)
toTreeArgs If        as = return (Node "If"            as)
toTreeArgs IfM       as = return (Node "IfM"           as)
toTreeArgs While     as = return (Node "While"         as)
toTreeArgs Locally   as = return (Node "Locally"       as)
toTreeArgs Embed     as = return (Node "Embed"         as)
toTreeArgs Return    as = return (Node "Return"        as)
toTreeArgs Bind      as = return (Node "Bind"          as)
toTreeArgs Mat       as = return (Node "Mat"           as)
toTreeArgs UnMat     as = return (Node "UnMat"         as)
toTreeArgs (Ops _ )  as = return (Node "ImageOperands" as)
toTreeArgs Coerce    as = return (Node "Coerce"        as)
toTreeArgs Undefined as = return (Node "Undefined"     as)
toTreeArgs NilHList  as = return (Node "NilHListAST"  as)
toTreeArgs ConsHList as = return (Node "ConsHListAST" as)
toTreeArgs HeadHList as = return (Node "HeadAST"      as)
toTreeArgs TailHList as = return (Node "TailAST"      as)
toTreeArgs ProductToHList   as = return (Node "ProductToHList"   as)
toTreeArgs (MkID     (v,_)) as = return (Node (show v) as)
toTreeArgs GradedMappend    as = return (Node "GradedMappend" as)
toTreeArgs (MkVector   n _) as = return (Node ("Vec"       ++ show (natVal n)) as)
toTreeArgs (Use    _ o    ) as = return (Node ("Use @("    ++ showSOptic o ++ ")") as)
toTreeArgs (Assign _ o    ) as = return (Node ("Assign @(" ++ showSOptic o ++ ")") as)
toTreeArgs (View   _ o    ) as = return (Node ("View @("   ++ showSOptic o ++ ")") as)
toTreeArgs (Set    _ o    ) as = return (Node ("Set @("    ++ showSOptic o ++ ")") as)
toTreeArgs (Def    k _    ) as = return (Node ("Def @"     ++ symbolVal k ) as)
toTreeArgs (FunDef k _ _  ) as = return (Node ("FunDef @"  ++ symbolVal k ) as)
toTreeArgs (Entry  _ (_ :: Proxy (stageInfo :: SPIRV.ExecutionInfo Nat stage) )) as
  = return (Node ("Entry @(" ++ show (knownValue @stage) ++ ")") as)
toTreeArgs (Lit (a :: ty)) as
  = return (Node ("Lit @(" ++ show (primTy @ty) ++ ") " ++ show a ) as)
toTreeArgs fm@Fmap as
  = case fm of
      ( _ :: AST ( (x -> y) -> f x -> f y ) )
        -> return (Node ("Fmap @(" ++ primFuncName @f ++ ") ") as)
toTreeArgs pur@Pure as
  = case pur of
      ( _ :: AST ( x -> f x ) )
        -> return (Node ("Pure @(" ++ primFuncName @f ++ ") ") as)
toTreeArgs app@Ap as
  = case app of
      ( _ :: AST ( f (x -> y) -> f x -> f y ) )
        -> return (Node ("Ap @("   ++ primFuncName @f ++ ") ") as)

toTree :: AST ast -> Tree String
toTree = (`evalState` (ID 1)) . runFreshSuccT . ( `toTreeArgs` [] )

instance Show (AST ast) where
  show = showTree . toTree
