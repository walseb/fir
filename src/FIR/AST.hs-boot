{-# LANGUAGE DataKinds    #-}
{-# LANGUAGE PolyKinds    #-}
{-# LANGUAGE TypeFamilies #-}

module FIR.AST
  ( AST, Code
  , Syntactic(..)
  , InternalType
  )
  where

-- haskus-utils-variant
import Haskus.Utils.EGADT
  ( EGADT )

-- fir
import FIR.AST.Type
  ( AugType(Val), UnderlyingType )
import {-# SOURCE #-} FIR.AST.ControlFlow
  ( IfF, IfMF, SwitchF, SwitchMF, WhileF )
import {-# SOURCE #-} FIR.AST.Effs
  ( LetF, DefF, FunDefF, FunCallF, DefEntryPointF, LocallyF, EmbedF )
import {-# SOURCE #-} FIR.AST.Exts
  ( DebugPrintfF, TraceRayF, ExecuteCallableF, RayQueryF )
import {-# SOURCE #-} FIR.AST.Images
  ( ImgOpsF, ImgQueryF )
import {-# SOURCE #-} FIR.AST.Optics
  ( UseF, AssignF, ViewF, SetF )
import {-# SOURCE #-} FIR.AST.Prim
  ( LamF, AppF, LitF, MkIDF, ValueF, UnsafeCoerceF
  , ReturnF, BindF
  , PrimOpF, UndefinedF, GradedMappendF
  , PureF, ApF
  , MkVectorF, MatF, UnMatF, StructF, ArrayF, ArrayLengthF, NilHListF, ConsHListF
  )

------------------------------------------------------------

type AllOpsF
  = '[ LamF, AppF, LitF, MkIDF, ValueF, UnsafeCoerceF
     , ReturnF, BindF
     , PrimOpF, UndefinedF, GradedMappendF
     , PureF, ApF
     , MkVectorF, MatF, UnMatF, StructF, ArrayF, ArrayLengthF, NilHListF, ConsHListF
     , IfF, IfMF, SwitchF, SwitchMF, WhileF
     , LetF, DefF, FunDefF, FunCallF, DefEntryPointF, LocallyF, EmbedF
     , DebugPrintfF, TraceRayF, ExecuteCallableF, RayQueryF
     , ImgOpsF, ImgQueryF
     , UseF, AssignF, ViewF, SetF
     ]

type AST    = EGADT AllOpsF
type Code a = AST (Val a)

class Syntactic a where
  type Internal a :: AugType
  toAST :: a -> AST (Internal a)
  fromAST :: AST (Internal a) -> a

type InternalType a = UnderlyingType (Internal a)
