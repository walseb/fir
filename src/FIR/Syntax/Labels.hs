{-# LANGUAGE AllowAmbiguousTypes    #-}
{-# LANGUAGE DataKinds              #-}
{-# LANGUAGE FlexibleContexts       #-}
{-# LANGUAGE FlexibleInstances      #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE GADTs                  #-}
{-# LANGUAGE PolyKinds              #-}
{-# LANGUAGE ScopedTypeVariables    #-}
{-# LANGUAGE TypeApplications       #-}
{-# LANGUAGE TypeOperators          #-}
{-# LANGUAGE UndecidableInstances   #-}

{-|
Module: FIR.Syntax.Labels
Description: @---@ __Optional__ /imperative-style/ syntax using overloaded labels.

Imperative-like syntax for variable assignment using overloaded labels. /Completely optional./

To use this functionality, import this module, activating /OverloadedLabels/ and /RebindableSyntax/.
You'll most likely want to use /TypeApplications/ too.

Usages of label syntax, such as @#foo@, will be desugared
using the 'fromLabel' method of the 'IsLabel' class in this module.

This allows for notation such as:

> do
>   #x #= (1 :: Float)  -- defines a new variable named 'x', initialised with value (1 :: Float)
> 
>   #x .= 2             -- sets the value of variable 'x' to 2
> 
>   #x %= (*3)          -- modifies the value of variable 'x' by multiplying it by 3
> 
>   #y @Float #=! 7     -- defines a new constant named 'y', initialised with value (7 :: Float)
> 
>   z <- #x + #y
>   -- 'z' then has value 6+7 = 13

In the first four examples, the overloaded labels are desugared into standalone labels,
which are then consumed by the infix operators @#=@, @.=@, @%=@ and @#=!@.

In the last example, the labels are desugared directly into values,
within the context of the indexed monad of the @do@ block.

The stateful nature of these labels is worth emphasising.
Consider for instance:

> do
>    #x @Float #= 0
>    x <- #x
>    #x .= 1
>    x' <- #x

When we write @x <- #x@, the /current/ value of @#x@ is obtained from the state.
In this example, this means that @x@ will continue to refer to the value @0@,
even after we set @#x@ to 1. When we retrieve the value of @#x@ after modifying it,
we then give @x'@ the (immutable) value of @1@.

This subtlety is especially important when working with @while@ loops,
as one wants to avoid having a constant conditional in the header.

-}

module FIR.Syntax.Labels
 ( -- * Overloaded labels
   IsLabel(fromLabel)
   -- * Infix operators
 , (#=), (#=!), (.=), (%=)
 )
 where

-- base
import Data.Kind
  ( Type )
import qualified GHC.Stack
import GHC.TypeLits
  ( Symbol, KnownSymbol )

-- fir
import Control.Monad.Indexed
  ( Codensity, (:=) )
import Control.Type.Optic
  ( Optic, Name )
import FIR.AST
  ( AST )
import FIR.Binding
  ( Var, R, RW )
import FIR.Syntax.Codensity
  ( def, use, assign, modifying )
import FIR.Prim.Singletons
  ( PrimTy )
import FIR.ProgramState
  ( ProgramState )
import FIR.Validation.Bindings
  ( ValidDef, AddBinding, Has, CanGet, CanPut )

--------------------------------------------------------------------------
-- overloaded labels

data Label (k :: Symbol) (a :: Type) = Label

data LabelUsage
  = Symbolic
  | Use Symbol ProgramState

class IsLabel k a (usage :: LabelUsage) v
    | v -> a, usage v -> k, v k -> usage
    where
  fromLabel :: GHC.Stack.HasCallStack => v

-- | Use a label symbolically.
--
-- That is, @#foo@ stands for a standalone label.
instance ( KnownSymbol k
         , PrimTy a
         , usage ~ Symbolic
         )
       => IsLabel k a usage (Label k a) where
  fromLabel = Label @k @a

-- | Use a label as a monadic value.
--
-- That is, @#foo@ stands for the value bound at @"foo"@
-- in the context of an indexed state monad.
instance ( KnownSymbol k
         , PrimTy a
         , a ~ Has k i
         , CanGet k i
         , r ~ (AST a := i)
         , usage ~ 'Use k i
         )
      => IsLabel k a usage (Codensity AST r i) where
  fromLabel = use @(Name k :: Optic '[] i a)

--------------------------------------------------------------------------
-- infix operators

infixr 1 #=
infixr 1 #=!
infixr 1 .=

-- | Define a new variable using a label.
(#=) :: forall a k i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , ValidDef k i
        , PrimTy a
        )
     => Label k a
     -> AST a
     -> Codensity AST (AST a := AddBinding k (Var RW a) i) i
_ #= a = def @k @RW a

-- | Define a new constant using a label.
(#=!) :: forall a k i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , ValidDef k i
        , PrimTy a
        )
     => Label k a
     -> AST a
     -> Codensity AST (AST a := AddBinding k (Var R a) i) i
_ #=! a = def @k @R a

-- | Set the value of a variable with given label.
(.=) :: forall a k i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , a ~ Has k i
        , CanPut k i
        , PrimTy a
        )
     => Label k a
     -> AST a
     -> Codensity AST (AST () := i) i
_ .= a = assign @(Name k :: Optic '[] i a) a

-- | Modify a variable with given label using a function.
(%=) :: forall a k i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , a ~ Has k i
        , CanGet k i
        , CanPut k i
        )
     => Label k a
     -> (AST a -> AST a)
     -> Codensity AST (AST () := i) i
_ %= f = modifying @(Name k :: Optic '[] i a) f
