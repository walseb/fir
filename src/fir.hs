{-# LANGUAGE AllowAmbiguousTypes  #-}
{-# LANGUAGE FlexibleContexts     #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE GADTs                #-}
{-# LANGUAGE OverloadedStrings    #-}
{-# LANGUAGE ScopedTypeVariables  #-}
{-# LANGUAGE TypeApplications     #-}
{-# LANGUAGE TypeOperators        #-}
{-# LANGUAGE UndecidableInstances #-}

module FIR 
  ( draw, compile
  , Arg(Debug, NoCode)
  , module Control.Monad.Indexed
  , Control.Type.Optic.Optic
  , (Control.Type.Optic.:*:)
  , (Control.Type.Optic.:.:)
  , Control.Type.Optic.Joint
  , Control.Type.Optic.AnIndex
  , Control.Type.Optic.Index
  , Control.Type.Optic.Name
  , Control.Type.Optic.ReifiedGetter(view)
  , Control.Type.Optic.ReifiedSetter(set)
  , (Data.Type.Map.:->)((:->))
  , FIR.AST.AST((:$))
  , FIR.AST.fromAST, FIR.AST.toAST -- might be a bad idea
  , FIR.Binding.BindingsMap
  , FIR.Binding.Var, FIR.Binding.Fun
  , FIR.Binding.R
  , FIR.Binding.W
  , FIR.Binding.RW
  , FIR.Definition.Definition
      ( Global
      , Function
      , EntryPoint
      )
  , FIR.Definition.Global_
  , FIR.Definition.Function_
  , FIR.Definition.EntryPoint_
  , module FIR.Instances.AST
  , module FIR.Instances.Codensity
  , module FIR.Instances.Optics
  , FIR.Prim.Array.Array
  , FIR.Prim.Array.mkArray
  , FIR.Prim.Array.RuntimeArray(MkRuntimeArray)
  , FIR.Prim.Struct.Struct(End,(:&))
  , FIR.Program.Procedure
  , FIR.Program.Program(Program)
  , module Math.Algebra.Class
  , module Math.Algebra.GradedSemigroup
  , module Math.Logic.Class
  , SPIRV.Decoration.Decoration
    -- explicit export list to avoid conflicting exports of 'Uniform'
    -- (conflicts with SPIRV.Storage.StorageClass(Uniform))
      ( SpecId, RowMajor, ColMajor
      , ArrayStride, MatrixStride
      , Builtin
      , NoPerspective, Flat
      , Patch
      , Stream, Location, Component
      , Index, Binding, DescriptorSet
      , Offset, Alignment
      )
  , SPIRV.ExecutionMode.ExecutionMode(..)
  , SPIRV.FunctionControl.Inlineability(..)
  , SPIRV.FunctionControl.SideEffects(..)
  , SPIRV.FunctionControl.FunctionControl
  , SPIRV.Stage.Stage(..)
  , SPIRV.Storage.StorageClass(Input, Output, Uniform, UniformConstant, PushConstant)
  ) where

-- base
import qualified Control.Monad as Monad (unless)
import Prelude

-- bytestring
import qualified Data.ByteString.Lazy as ByteString

-- tree-view
import Data.Tree.View(drawTree)

-- text-utf8
import Data.Text(Text)

-- fir
import CodeGen.CodeGen
import CodeGen.State
import Control.Monad.Indexed
import Control.Type.Optic
import Data.Type.Map
import FIR.AST
import FIR.Binding
import FIR.Definition
import FIR.Instances.AST
import FIR.Instances.Codensity
import FIR.Instances.Optics
import FIR.Prim.Array
import FIR.Prim.Struct
import FIR.Program
import Math.Algebra.Class
import Math.Algebra.GradedSemigroup
import Math.Logic.Class
import SPIRV.Decoration
import SPIRV.ExecutionMode
import SPIRV.FunctionControl
import SPIRV.Stage
import SPIRV.Storage

------------------------------------------------

class DrawableProgram prog where
  draw :: prog -> IO ()

instance DrawableProgram (Codensity AST (AST a := j) i) where
  draw = drawTree . toTree . toAST
instance ( DrawableProgram
             ( UndecoratedProgram (InsertionSort defs) a )
         )
      => DrawableProgram (Program defs a)
      where
  draw (Program prog) = draw prog


data Arg
  = NoCode
  | Debug
  deriving ( Prelude.Eq, Show )

compile :: forall defs a. KnownDefinitions defs
        => FilePath
        -> [Arg]
        -> Program defs a
        -> IO ( Either Text Text )
compile filePath args (Program program) = case runCodeGen cgContext (toAST program) of
    Left  err -> Prelude.pure ( Left err )
    Right bin
      ->  do  Monad.unless ( NoCode `elem` args )
                ( ByteString.writeFile filePath bin )
              Prelude.pure ( Right "OK" )
  where cgContext :: CGContext
        cgContext = (context @defs) { debugMode = Debug `elem` args }