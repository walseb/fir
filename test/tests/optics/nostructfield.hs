{-# LANGUAGE BlockArguments      #-}
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE FlexibleInstances   #-}
{-# LANGUAGE GADTs               #-}
{-# LANGUAGE OverloadedLabels    #-}
{-# LANGUAGE PatternSynonyms     #-}
{-# LANGUAGE PolyKinds           #-}
{-# LANGUAGE RebindableSyntax    #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications    #-}
{-# LANGUAGE TypeOperators       #-}

module Tests.Optics.NoStructField where

-- base
import Prelude hiding ( Functor(..), (<$>)
                      , Applicative(..), Monad(..)
                      , Num(..), Fractional(..), Integral(..), Floating(..)
                      , Eq(..), Ord(..)
                      , (&&)
                      )
import qualified Prelude

-- vector
import qualified Data.Vector as Array

-- fir
import FIR
import FIR.Labels
import Math.Linear

------------------------------------------------
-- program

type Defs
  = '[ "vertexData" ':-> Global_ Input
          ( Struct [ "position" ':-> V 3 Float
                    , "colour"   ':-> V 4 Float
                    , "size"     ':-> Float
                    , "weight"   ':-> Float
                    ]
          )
     ]

program :: Program Defs ()
program = Program do

  entryPoint @"main" @Vertex do

    position <- use @( Name "vertexData" :.: Name "inexistent" )

    #gl_Position .= position