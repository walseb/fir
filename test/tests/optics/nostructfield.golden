
test\tests\optics\nostructfield.hs:50:17: error:
    • get: struct has no field with name "inexistent".
      This struct has the following fields:
      '[ "position" ':-> V 3 Float, "colour" ':-> V 4 Float,
         "size" ':-> Float, "weight" ':-> Float]
    • In a stmt of a 'do' block:
        position <- use @(Name "vertexData" :.: Name "inexistent")
      In the third argument of ‘entryPoint’, namely
        ‘do position <- use @(Name "vertexData" :.: Name "inexistent")
            #gl_Position .= position’
      In a stmt of a 'do' block:
        entryPoint
          @"main"
          @Vertex
          do position <- use @(Name "vertexData" :.: Name "inexistent")
             #gl_Position .= position
   |
50 |     position <- use @( Name "vertexData" :.: Name "inexistent" )
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
